cypress-io/cypress
# feat: Add banners for incomplete project connection states

- Closes #22637 

Apologies for the size of this - 2/3 is new tests, so the actual logic changes are fairly minimal. Mostly moving stuff into new shared components.

### User facing changelog
* Display a banner with a corrective action when your project uses a `projectId` that cannot be found in the Cypress Dashboard
* Display a banner with a corrective action when you do not have access to the current project in the Cypress Dashboard

### Additional details
* Refactoring existing "Request Access" button logic into a new shared component
* Extract existing banners in `SpecsList` to a new `SpecsListBanners` component
* Remove manual tracking of "hasRequestedAccess" in favor of updating GQL cache with mutation result - this allows the correct state to be shared across the app
* Update GraphQLDataSource to allow the special `CloudProjectUnauthorized` and `CloudProjectNotFound` types to be propagated to the frontend so they display correctly on first load

### Steps to test
1. Open `cypress-example-kitchensink` in the app, ensure you are logged in
2. In the e2e Specs Explorer, verify that Latest Runs data is loading
3. Open the `cypress.config.js` file for the project and replace the `projectId` with `abc123`
4. Close and reopen the project. Verify that a banner is displayed indicating the invalid project ID and providing a button to reconnect the project
5. Open the `cypress.config.js` file for the project and replace the `projectId` with `zqgm7c`
6. Close and reopen the project. Verify that a banner is displayed indicating you are not authorized to view the specified project and providing a button to request access. **Do not request access yet**
7. Mouse over the "Latest Runs" and "Average Duration" headers, verify the displayed tooltips also provide "Request Access" buttons
8. Click any one of the three "Request Access" buttons you've seen so far. Verify it updates to disabled with a "Request Sent" label
9. Verify the other two "Request Access" buttons have also updated and that the banner text has updated to reflect the submission has been made

### How has the user experience changed?
**Bad project ID (not found)**

New banner should display at top of specs list, should provide same action as header tooltips to reconnect project

![Screen Shot 2022-07-18 at 2 06 19 PM](https://user-images.githubusercontent.com/11482842/179592116-bf5b2716-8f13-4b58-980e-b56930562c5d.png)


**Unauthorized (request access)**

New banner should display at top of specs list, should provide same action as header tooltips to request access. Button and content should update to reflect a pending access request.

https://user-images.githubusercontent.com/11482842/179592333-8e2189ef-47d6-48c6-8ab8-c57ff8f6b216.mov


### PR Tasks

- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
